File: Post-segmentation analysis/health utilization score/lagos_health_utilization/lagos_health_utilization.R
rm(list=ls())
library(haven)
library(ggplot2)
library(gridExtra)
library(reshape)
library(survival)
library(magrittr)
library(dplyr)
library(tidyr)
library(data.table)
library(survey)
library(plyr)
library(tm)
library(labelled)
library(stringr)
library(forcats)
require(doParallel)
library(foreign)
library(psych)
library(vtable)
library(corrplot)
library(haven)
library(survey)

############## Load Nigeria 2013 DHS data
NGHH <- data.table(read.dta("data/NGHR6ADT/NGHR6AFL.DTA"))
NGIR <- data.table(read.dta("data/NGIR6ADT/NGIR6AFL.DTA"))

############## read Pathways Rdat file
path <- load("clean_data_files/DHS Pathways interim segments & factors.RDat")

dhs.seg <- data.table(dhs.seg)
unique(dhs.seg$cv.regX5)

# Select out only the Nigeria segments
ng.seg <- dhs.seg[cv.regX5 %in% c("Nigeria.South"),]

# We need to re-create a caseid that can be linked to the DHS file, this requires parsing out cluster,
# HHID and Lineno separately, the putting them together with a specific number of characters
ng.seg$ID<-as.character(ng.seg$IDHSPID)
#str_length(ng.seg$IDHSPID) #23
ng.seg$cluster<-str_sub(ng.seg$ID,10,17)
ng.seg$hhid<-str_sub(ng.seg$ID,18,20)
ng.seg$hhid<-str_pad(ng.seg$hhid,3)
ng.seg$lineno<-str_sub(ng.seg$ID,21,23)
ng.seg$lineno<-str_pad(ng.seg$lineno,2)
ng.seg$caseid<-paste(ng.seg$cluster,ng.seg$hhid,ng.seg$lineno, sep = "")
ng.seg$caseid<-str_trim(ng.seg$caseid,"left")

ng.seg<- ng.seg %>% dplyr::select(caseid,segX4)

################################################### Define zero dose child
NGBR <- data.table(read.dta("data/NGBR6ADT/NGBR6AFL.DTA"))

table(NGBR$h3,useNA="always")
table(NGBR$h5,useNA="always")
table(NGBR$h7,useNA="always")

NGBR[!(is.na(h3)),dpt1:=ifelse(h3 %in% c(1,2,3),1,0)]  # lacked or received the first dose of dpt
NGBR[!(is.na(h5)),dpt2:=ifelse(h5 %in% c(1,2,3),1,0)]  # lacked or received the second dose of dpt
NGBR[!(is.na(h7)),dpt3:=ifelse(h7 %in% c(1,2,3),1,0)]  # lacked or received the third dose of dpt

###################################################
####### for this analysis, similar to Roy's analysis, I defined zero dose as dpt0 means no dose of dpt for the 10-23 motnhs old children
###################################################

NGBR[,dpt0:= ifelse((dpt1+dpt2+dpt3)==0, 1, 0)] #no dose of dpt
table(NGBR$dpt0,useNA="always")
table(NGBR$v008,useNA="always")
table(NGBR$b3,useNA="always")

NGBR[,ChildAge:= (v008-b3)] # in months
table(NGBR$b5,useNA="always")
NGBR <- NGBR[b5=="Yes",] #child is alive

NGBR$Agegrp <- ifelse(NGBR$ChildAge < 12, "<12",ifelse(NGBR$ChildAge >=12 & NGBR$ChildAge <=23, "12-23",ifelse(NGBR$ChildAge>23 &NGBR$ChildAge<=35,"24-35",
ifelse(NGBR$ChildAge>35 &NGBR$ChildAge<=59,"36-59","60+"))))

NGBR$Agegrp <- factor(NGBR$Agegrp, levels = c("<12","12-23", "24-35","36-59","60+"))

NGBR[,table(dpt0,Agegrp,useNA="always")]

NGBR$dpt0 <- ifelse(NGBR$ChildAge>9 & NGBR$ChildAge <24,NGBR$dpt0,NA)

###################################################
# Now we can link back to any of the Nigeria DHS files
Merg <- ng.seg %>% left_join(NGBR)

## ## ## ##
## ## ## ## Seek care for diarrhea and fever
## ## ## ##
table(NGBR$h12z,useNA="always")
table(NGBR$h32z ,useNA="always")

## diarrhea: medical treatment 0:No, 1:Yes
NGBR$diarr <- NGBR$h12z
## fever: medical treatment 0:No, 1:Yes
NGBR$fever <- NGBR$h32z

NGBR$diarr <- ifelse(NGBR$diarr== 9,NA,NGBR$diarr)
NGBR$fever <- ifelse(NGBR$fever== 9,NA,NGBR$fever)
table(NGBR$diarr,useNA="always")
table(NGBR$fever,useNA="always")

diar.fev <- NGBR[,list(diarr=max(diarr,na.rm=TRUE),fever=max(fever,na.rm=TRUE)),by=c("caseid")]
diar.fev$diarr <-  ifelse(diar.fev$diarr==-Inf,NA,diar.fev$diarr)
diar.fev$fever <-  ifelse(diar.fev$fever==-Inf,NA,diar.fev$fever)

table(diar.fev$fever ,useNA="always")
table(diar.fev$diarr ,useNA="always")

###################################################
# Now we can link back to any of the Nigeria DHS files
Merg <- Merg %>% left_join(diar.fev)

# Finally we need to apply the new coding for the segments
# The segment renaming come from a slide sent by Melanie specifically for the Nigeria DHS data
old<-c("RF1","RF2","RF3","RF4","RM1","RM2","RM3","UF1","UF2","UF3","UM1","UM2","UM3","UM4","UM5") # What's in the file
new<-c("RF4.2","RF3","RF4.1","RF1","RM2.1","RM4","RM2.2","UF3.1","UF3.2","UF1","UM4","UM2","UM1.1","UM3","UM1.2") # New segments
comb<-c("RF4","RF3","RF4","RF1","RM2","RM4","RM3","UF3","UF3","UF1","UM4","UM2","UM1","UM3","UM1") # Combining to be more like Pathways
new.seg<-data.frame("segX4"=old, "new.seg"=new, "comb.seg"=comb)

Merg<-Merg%>%left_join(new.seg)

# Just use the segments that also appear in the Pathways data
Merg <- Merg %>% filter(new.seg %in% c("UM1.1","UM1.2","UF1","UM3","UF3.1","UM4"))

nigeria <- Merg
#nigeria$segm1 <- nigeria$comb.seg
nigeria$segm1 <- nigeria$new.seg

table(nigeria$segm1,nigeria$dpt0,useNA="always")
table(nigeria$v024,useNA="always")


#nigeria$region <- ifelse(nigeria$v024 %in% c("North Central","North East","North West"),"north","south")
nigeria$wt<-nigeria$v005/1000000


## ## ## ## ## ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## ##
## ## ## ## ## ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## ##
## ## ## ## SCORING
## ## ## ## ## ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## ##
## ## ## ## ## ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## ##

### anc
table(nigeria$m14,useNA="always")
nigeria$anyanc <- ifelse(nigeria$m14==99,NA,ifelse(nigeria$m14==0,0,1))
table(nigeria$anyanc,useNA="always")

nigeria$anc.num2 <- ifelse(nigeria$m14 == 98,3,nigeria$m14) # Impute DNK to the mean of 3
nigeria$anc.num2 <- ifelse(nigeria$anc.num2== 99,NA,nigeria$anc.num2)
table(nigeria$anc.num2,useNA="always")

nigeria$anc.4plus2 <-ifelse(nigeria$anc.num>=4,1,0)
table(nigeria$anc.4plus2,useNA="always")

##### PNC
table(nigeria$m70,useNA="always")
nigeria[!(is.na(m70)),B.PNC := ifelse(m70==1,1,0)]
nigeria$B.PNC <- ifelse(nigeria$m70== 9,NA,nigeria$B.PNC)
table(nigeria$B.PNC,useNA="always")

# Recode FP
table(nigeria$v361 ,useNA="always")
nigeria$fp.ever <- ifelse(nigeria$v361 == "Never used",0,1)

# Recode used health services
table(nigeria$v393 ,useNA="always")
table(nigeria$v394 ,useNA="always")
nigeria$hc.chw <- ifelse(nigeria$v393 == 1,1,0)
nigeria$hc.chw <- ifelse(nigeria$v393 == 9,NA,nigeria$hc.chw)

nigeria$hc.hfac <- ifelse(nigeria$v394 == 1,1,0)
nigeria$hc.hfac <- ifelse(nigeria$v394 == 9,NA,nigeria$hc.hfac)
table(nigeria$hc.hfac,useNA="always")

# Recode home births
table(NGIR$m15_1,useNA="always")
NGIR[,hb_1:= ifelse(m15_1== 10 | m15_1== 11 | m15_1== 12 | m15_1== 96,1,0)]
NGIR[,hb_2:= ifelse(m15_2== 10 | m15_2== 11 | m15_2== 12 | m15_2== 96,1,0)]
NGIR[,hb_3:= ifelse(m15_3== 10 | m15_3== 11 | m15_3== 12 | m15_3== 96,1,0)]
NGIR[,hb_4:= ifelse(m15_4== 10 | m15_4== 11 | m15_4== 12 | m15_4== 96,1,0)]
NGIR[,hb_5:= ifelse(m15_5== 10 | m15_5== 11 | m15_5== 12 | m15_5== 96,1,0)]

NGIR$hb_1 <- ifelse(NGIR$m15_1 == 99,NA,NGIR$hb_1)
NGIR$hb_2 <- ifelse(NGIR$m15_2 == 99,NA,NGIR$hb_2)
NGIR$hb_3 <- ifelse(NGIR$m15_3 == 99,NA,NGIR$hb_3)
NGIR$hb_4 <- ifelse(NGIR$m15_4 == 99,NA,NGIR$hb_4)
NGIR$hb_5 <- ifelse(NGIR$m15_5 == 99,NA,NGIR$hb_5)

NGIR$hb_ever[(NGIR$hb_1==1 | NGIR$hb_2==1 | NGIR$hb_3==1 | NGIR$hb_4==1 | NGIR$hb_5==1)]<- 1
NGIR$hb_ever[((NGIR$hb_1==0 | is.na(NGIR$hb_1)) & (NGIR$hb_2==0 | is.na(NGIR$hb_2)) & (NGIR$hb_3==0 | is.na(NGIR$hb_3)) &
                (NGIR$hb_4==0 | is.na(NGIR$hb_4)) & (NGIR$hb_5==0 | is.na(NGIR$hb_5)))]<-0
NGIR$hb_ever <- ifelse((is.na(NGIR$hb_1))&(is.na(NGIR$hb_2))&(is.na(NGIR$hb_3))&(is.na(NGIR$hb_4))&(is.na(NGIR$hb_5)),NA,NGIR$hb_ever)

NGIR[,table(hb_ever, useNA="always")]
NGIR[,table(hb_1, useNA="always")]

NGIR$caseid<-str_trim(NGIR$caseid,"left")

nigeria <- merge(nigeria, NGIR[,c("caseid","hb_ever")], by=c("caseid"))

## ## ## ## ## ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## #### ## ## ##
table(nigeria$anyanc,useNA="always")
table(nigeria$hc.hfac,useNA="always")
table(nigeria$B.PNC,useNA="always")
table(nigeria$fp.ever,useNA="always")
table(nigeria$hb_ever,useNA="always")
table(nigeria$fever,useNA="always")
table(nigeria$diarr,useNA="always")

### woman & child health service utilization
nigeria[,HC.score4:= (anyanc+hc.hfac+B.PNC+fp.ever+diarr+fever)]
nigeria[hb_ever==0,HC.score4:= HC.score4 + 1]
nigeria[is.na(hb_ever),HC.score4:= NA]

table(nigeria$HC.score4,useNA="always")
table(nigeria[is.na(hb_ever),]$HC.score4,useNA="always")

######
nigeria2 <- nigeria[!duplicated(nigeria$caseid), ]
table(nigeria2$segm1,useNA="always")

table(nigeria2$segm1,nigeria2$anyanc,useNA="always")
table(nigeria2$segm1,nigeria2$hc.hfac,useNA="always")
table(nigeria2$segm1,nigeria2$B.PNC,useNA="always")
table(nigeria2$segm1,nigeria2$fp.ever,useNA="always")
table(nigeria2$segm1,nigeria2$diarr,useNA="always")
table(nigeria2$segm1,nigeria2$fever,useNA="always")
table(nigeria2$segm1,nigeria2$hb_ever,useNA="always")

nigeria2$anyanc <- ifelse(is.na(nigeria2$anyanc),"NA",nigeria2$anyanc)
nigeria2$hc.hfac <- ifelse(is.na(nigeria2$hc.hfac),"NA",nigeria2$hc.hfac)
nigeria2$B.PNC <- ifelse(is.na(nigeria2$B.PNC),"NA",nigeria2$B.PNC)
nigeria2$fp.ever <- ifelse(is.na(nigeria2$fp.ever),"NA",nigeria2$fp.ever)
nigeria2$diarr <- ifelse(is.na(nigeria2$diarr),"NA",nigeria2$diarr)
nigeria2$fever <- ifelse(is.na(nigeria2$fever),"NA",nigeria2$fever)
nigeria2$hb_ever <- ifelse(is.na(nigeria2$hb_ever),"NA",nigeria2$hb_ever)

table(nigeria2$hb_ever,useNA="always")
nigeria2$hf_deliv <- ifelse(nigeria2$hb_ever==0,1,ifelse(nigeria2$hb_ever==1,0,NA))
nigeria2$hf_deliv <- ifelse(is.na(nigeria2$hf_deliv),"NA",nigeria2$hf_deliv)
table(nigeria2$hf_deliv,useNA="always")


nigeria2 <- nigeria2 %>% filter(!is.na(v021))

Pathdesign<-svydesign(id=nigeria2$v021, strata= nigeria2$v023, weights=nigeria2$wt, survey.lonely.psu="adjust", nest=T, data=nigeria2)

# get totals that are represented by segment
Pathdesign<-svydesign(id=nigeria2$v021, weights=nigeria2$wt, survey.lonely.psu="adjust", nest=T, data=nigeria2)
totals <- svytotal(~segm1, design = Pathdesign)



tot <- melt(svytable(~segm1, Pathdesign))

tot <-data.table(tot)
setkey(tot,value)

# Function to compute subtracted value.y for each segment group

######## anc
seg <- melt(svytable(~segm1+anyanc, Pathdesign))
anc <- merge(seg,tot,by="segm1")

subtract_value_y <- function(data) {
  if (any(is.na(data$anyanc))) {
    na_row_value_x <- data$value.x[which(is.na(data$anyanc))]
    data$value.y <- data$value.y - na_row_value_x
  }
  return(data)
}

anc <- anc %>%
  group_by(segm1) %>%
  do(subtract_value_y(.)) %>%
  ungroup()

anc$perct <- 100*anc$value.x/anc$value.y
anc$health.service <- "anc"
names(anc)[2] <-"health.service.status"

######## visit health facility
seg <- melt(svytable(~segm1+hc.hfac, Pathdesign))
hc.hfac <- merge(seg,tot,by="segm1")

subtract_value_y <- function(data) {
  if (any(is.na(data$hc.hfac))) {
    na_row_value_x <- data$value.x[which(is.na(data$hc.hfac))]
    data$value.y <- data$value.y - na_row_value_x
  }
  return(data)
}

hc.hfac <- hc.hfac %>%
  group_by(segm1) %>%
  do(subtract_value_y(.)) %>%
  ungroup()

hc.hfac$perct <- 100*hc.hfac$value.x/hc.hfac$value.y
hc.hfac$health.service <- "hc.hfac"
names(hc.hfac)[2] <-"health.service.status"

######## B.PNC
seg <- melt(svytable(~segm1+B.PNC, Pathdesign))
B.PNC <- merge(seg,tot,by="segm1")

subtract_value_y <- function(data) {
  if (any(is.na(data$B.PNC))) {
    na_row_value_x <- data$value.x[which(is.na(data$B.PNC))]
    data$value.y <- data$value.y - na_row_value_x
  }
  return(data)
}

B.PNC <- B.PNC %>%
  group_by(segm1) %>%
  do(subtract_value_y(.)) %>%
  ungroup()


B.PNC$perct <- 100*B.PNC$value.x/B.PNC$value.y
B.PNC$health.service <- "B.PNC"
names(B.PNC)[2] <-"health.service.status"

######## ever used FP
seg <- melt(svytable(~segm1+fp.ever, Pathdesign))
fp.ever <- merge(seg,tot,by="segm1")

fp.ever$perct <- 100*fp.ever$value.x/fp.ever$value.y
fp.ever$health.service <- "fp.ever"
names(fp.ever)[2] <-"health.service.status"

######## diarrhea
seg <- melt(svytable(~segm1+diarr, Pathdesign))
diarr <- merge(seg,tot,by="segm1")

subtract_value_y <- function(data) {
  if (any(is.na(data$diarr))) {
    na_row_value_x <- data$value.x[which(is.na(data$diarr))]
    data$value.y <- data$value.y - na_row_value_x
  }
  return(data)
}

diarr <- diarr %>%
  group_by(segm1) %>%
  do(subtract_value_y(.)) %>%
  ungroup()

diarr$perct <- 100*diarr$value.x/diarr$value.y
diarr$health.service <- "diarr"
names(diarr)[2] <-"health.service.status"

######## fever
seg <- melt(svytable(~segm1+fever, Pathdesign))
fever <- merge(seg,tot,by="segm1")

subtract_value_y <- function(data) {
  if (any(is.na(data$fever))) {
    na_row_value_x <- data$value.x[which(is.na(data$fever))]
    data$value.y <- data$value.y - na_row_value_x
  }
  return(data)
}

fever <- fever %>%
  group_by(segm1) %>%
  do(subtract_value_y(.)) %>%
  ungroup()

fever$perct <- 100*fever$value.x/fever$value.y
fever$health.service <- "fever"
names(fever)[2] <-"health.service.status"

######## facility delivery
seg <- melt(svytable(~segm1+hf_deliv, Pathdesign))
hf_deliv <- merge(seg,tot,by="segm1")

subtract_value_y <- function(data) {
  if (any(is.na(data$hf_deliv))) {
    na_row_value_x <- data$value.x[which(is.na(data$hf_deliv))]
    data$value.y <- data$value.y - na_row_value_x
  }
  return(data)
}

hf_deliv <- hf_deliv %>%
  group_by(segm1) %>%
  do(subtract_value_y(.)) %>%
  ungroup()

hf_deliv$perct <- 100*hf_deliv$value.x/hf_deliv$value.y
hf_deliv$health.service <- "hf_deliv"
names(hf_deliv)[2] <-"health.service.status"

health.service.lagos <- rbind(anc,hc.hfac,B.PNC,fp.ever,diarr,fever,hf_deliv)
names(health.service.lagos)[5] <- "percentage"
names(health.service.lagos)[4] <- "total_in_segment"
names(health.service.lagos)[3] <- "number"

health.service.lagos <- health.service.lagos %>% filter(!is.na(health.service.status))
health.service.lagos$segm1 <- as.character(health.service.lagos$segm1)
health.service.lagos$health.service.status <- as.character(health.service.lagos$health.service.status)

health.service.lagos <- as.data.frame(health.service.lagos)

# create averages for utilizaton

health.service.lagos.ave <- health.service.lagos %>%
  dplyr::group_by(segm1, health.service.status) %>%
  dplyr::summarise(average.utilization = mean(percentage)) %>%
  filter(health.service.status == 1)

write.csv(health.service.lagos,"heath_utilization_by_service_lagos.csv")
write.csv(health.service.lagos.ave,"heath_utilization_average_lagos.csv")


#

File: Post-segmentation analysis/health utilization score/lagos_health_utilization/lagos_health_utilization.py
import pandas as pd
import numpy as np
import rpy2.robjects as robjects
from rpy2.robjects import pandas2ri
from rpy2.robjects.packages import importr
import os

# Activate the pandas to R interface
pandas2ri.activate()

# Import R libraries
base = importr('base')
stats = importr('stats')
survey = importr('survey')
haven = importr('haven')
foreign = importr('foreign')

# Set working directory to the script's location
# os.chdir(os.path.dirname(os.path.abspath(__file__)))

############## Load Nigeria 2013 DHS data
# Using pandas to read Stata files (DTA)
nghh = pd.read_stata("data/NGHR6ADT/NGHR6AFL.DTA", convert_categoricals=False)
ngir = pd.read_stata("data/NGIR6ADT/NGIR6AFL.DTA", convert_categoricals=False)

############## Read Pathways Rdat file
# We use rpy2 to load the RData file and bring it to pandas
robjects.r['load']("clean_data_files/DHS Pathways interim segments & factors.RDat")
dhs_seg_r = robjects.r['dhs.seg']
dhs_seg = pandas2ri.rpy2py(dhs_seg_r)

# Select out only the Nigeria segments
ng_seg = dhs_seg[dhs_seg['cv.regX5'] == "Nigeria.South"].copy()

# Re-create caseid
ng_seg['ID'] = ng_seg['IDHSPID'].astype(str)
ng_seg['cluster'] = ng_seg['ID'].str[9:17]
ng_seg['hhid'] = ng_seg['ID'].str[17:20].str.rjust(3)
ng_seg['lineno'] = ng_seg['ID'].str[20:23].str.rjust(2)
ng_seg['caseid'] = (ng_seg['cluster'] + ng_seg['hhid'] + ng_seg['lineno']).str.strip()

ng_seg = ng_seg[['caseid', 'segX4']]

################################################### Define zero dose child
ngbr = pd.read_stata("data/NGBR6ADT/NGBR6AFL.DTA", convert_categoricals=False)

def get_dpt(val):
    if pd.isna(val):
        return np.nan
    return 1 if val in [1, 2, 3] else 0

ngbr['dpt1'] = ngbr['h3'].apply(get_dpt)
ngbr['dpt2'] = ngbr['h5'].apply(get_dpt)
ngbr['dpt3'] = ngbr['h7'].apply(get_dpt)

ngbr['dpt0'] = ((ngbr['dpt1'] + ngbr['dpt2'] + ngbr['dpt3']) == 0).astype(int)
ngbr['ChildAge'] = ngbr['v008'] - ngbr['b3']
ngbr = ngbr[ngbr['b5'] == 1].copy() # b5 == 1 is "Yes" usually in DHS DTA

# dpt0 is for 10-23 months
ngbr['dpt0'] = np.where((ngbr['ChildAge'] > 9) & (ngbr['ChildAge'] < 24), ngbr['dpt0'], np.nan)

################################################### Seek care for diarrhea and fever
ngbr['diarr'] = ngbr['h12z'].replace(9, np.nan)
ngbr['fever'] = ngbr['h32z'].replace(9, np.nan)

diar_fev = ngbr.groupby('caseid').agg({
    'diarr': lambda x: x.max() if x.notna().any() else np.nan,
    'fever': lambda x: x.max() if x.notna().any() else np.nan
}).reset_index()

################################################### Link back
merg = ng_seg.merge(ngbr, on='caseid', how='left')
merg = merg.merge(diar_fev, on='caseid', how='left')

# Segment renaming
segment_map = pd.DataFrame({
    'segX4': ["RF1","RF2","RF3","RF4","RM1","RM2","RM3","UF1","UF2","UF3","UM1","UM2","UM3","UM4","UM5"],
    'new_seg': ["RF4.2","RF3","RF4.1","RF1","RM2.1","RM4","RM2.2","UF3.1","UF3.2","UF1","UM4","UM2","UM1.1","UM3","UM1.2"],
    'comb_seg': ["RF4","RF3","RF4","RF1","RM2","RM4","RM3","UF3","UF3","UF1","UM4","UM2","UM1","UM3","UM1"]
})

merg = merg.merge(segment_map, on='segX4', how='left')
merg = merg[merg['new_seg'].isin(["UM1.1","UM1.2","UF1","UM3","UF3.1","UM4"])].copy()

nigeria = merg.copy()
nigeria['segm1'] = nigeria['new_seg']
nigeria['wt'] = nigeria['v005'] / 1000000

## SCORING
nigeria['anyanc'] = np.where(nigeria['m14'] == 99, np.nan, np.where(nigeria['m14'] == 0, 0, 1))
# anc.num2 and anc.4plus2 are calculated but not used in final score calculation in R script except for HC.score4
# But HC.score4 is not used in the final summary.

# PNC
nigeria['B_PNC'] = np.where(nigeria['m70'] == 9, np.nan, np.where(nigeria['m70'] == 1, 1, 0))

# FP
nigeria['fp_ever'] = np.where(nigeria['v361'] == 0, 0, 1) # Assuming 0 is "Never used"

# Health services
nigeria['hc_chw'] = np.where(nigeria['v393'] == 9, np.nan, np.where(nigeria['v393'] == 1, 1, 0))
nigeria['hc_hfac'] = np.where(nigeria['v394'] == 9, np.nan, np.where(nigeria['v394'] == 1, 1, 0))

# Home births in ngir
for i in range(1, 6):
    col = f'm15_{i}'
    hb_col = f'hb_{i}'
    ngir[hb_col] = np.where(ngir[col].isin([10, 11, 12, 96]), 1, 0)
    ngir[hb_col] = np.where(ngir[col] == 99, np.nan, ngir[hb_col])

ngir['hb_ever'] = ngir[[f'hb_{i}' for i in range(1, 6)]].max(axis=1)
ngir['caseid'] = ngir['caseid'].str.strip()

nigeria = nigeria.merge(ngir[['caseid', 'hb_ever']], on='caseid', how='left')

# final indicators for summary
nigeria['hf_deliv'] = np.where(nigeria['hb_ever'] == 0, 1, np.where(nigeria['hb_ever'] == 1, 0, np.nan))

# Deduplicate by caseid for survey analysis as in R script
nigeria2 = nigeria.drop_duplicates(subset='caseid').copy()
nigeria2 = nigeria2[nigeria2['v021'].notna()]

# Convert to R data frame for survey package
r_nigeria2 = pandas2ri.py2rpy(nigeria2)
robjects.globalenv['nigeria2'] = r_nigeria2

# Execute survey design and tables in R via rpy2
robjects.r('''
    library(survey)
    library(reshape2)
    library(dplyr)
    
    Pathdesign <- svydesign(id=~v021, weights=~wt, data=nigeria2)
    
    tot <- as.data.frame(svytable(~segm1, Pathdesign))
    colnames(tot) <- c("segm1", "value.y")
    
    summarize_service <- function(var_name, service_label) {
        formula_str <- paste0("~segm1 + ", var_name)
        seg <- as.data.frame(svytable(as.formula(formula_str), Pathdesign))
        colnames(seg) <- c("segm1", "health.service.status", "value.x")
        
        # Merge with totals
        m <- merge(seg, tot, by="segm1")
        
        # Subtract NAs if present
        na_rows <- m[is.na(m$health.service.status), ]
        if (nrow(na_rows) > 0) {
            for (s in unique(m$segm1)) {
                na_val <- na_rows$value.x[na_rows$segm1 == s]
                if (length(na_val) > 0) {
                    m$value.y[m$segm1 == s] <- m$value.y[m$segm1 == s] - na_val
                }
            }
        }
        
        m <- m[!is.na(m$health.service.status), ]
        m$perct <- 100 * m$value.x / m$value.y
        m$health.service <- service_label
        return(m)
    }
    
    anc <- summarize_service("anyanc", "anc")
    hc_hfac <- summarize_service("hc_hfac", "hc_hfac")
    B_PNC <- summarize_service("B_PNC", "B.PNC")
    fp_ever <- summarize_service("fp_ever", "fp.ever")
    diarr <- summarize_service("diarr", "diarr")
    fever <- summarize_service("fever", "fever")
    hf_deliv <- summarize_service("hf_deliv", "hf_deliv")
    
    health_service_lagos <- rbind(anc, hc_hfac, B_PNC, fp_ever, diarr, fever, hf_deliv)
    colnames(health_service_lagos)[colnames(health_service_lagos) == "value.x"] <- "number"
    colnames(health_service_lagos)[colnames(health_service_lagos) == "value.y"] <- "total_in_segment"
    colnames(health_service_lagos)[colnames(health_service_lagos) == "perct"] <- "percentage"
''')

health_service_lagos = pandas2ri.rpy2py(robjects.r['health_service_lagos'])

# Calculate averages as in R script
health_service_lagos_ave = health_service_lagos[health_service_lagos['health.service.status'] == "1"].copy()
health_service_lagos_ave['percentage'] = pd.to_numeric(health_service_lagos_ave['percentage'])
health_service_lagos_ave = health_service_lagos_ave.groupby(['segm1', 'health.service.status'])['percentage'].mean().reset_index()
health_service_lagos_ave.rename(columns={'percentage': 'average.utilization'}, inplace=True)

# Write outputs
health_service_lagos.to_csv("heath_utilization_by_service_lagos.csv", index=False)
health_service_lagos_ave.to_csv("heath_utilization_average_lagos.csv", index=False)
